Userland/Libraries/LibWeb/Painting/StackingContext.cpp

```diff
@@ -146,61 +154,55 @@ void StackingContext::paint_internal(PaintContext& context) const
             paintable->after_children_paint(context, PaintPhase::Foreground, should_clip_overflow);
     };
 
+    // For a more elaborate description of the algorithm, see CSS 2.1 Appendix E
+    // Draw the background and borders for the context root (steps 1, 2)
+    paint_box_background(m_box);
+
     // Draw positioned descendants with negative z-indices (step 3)
     for (auto* child : m_children) {
         if (child->m_box.computed_values().z_index().has_value() && child->m_box.computed_values().z_index().value() < 0)
             paint_child(child);
     }
 
-    // Draw the background and borders for block-level children (step 4)
-    paint_descendants(context, m_box, StackingContextPaintPhase::BackgroundAndBorders);
-    // Draw the non-positioned floats (step 5)
-    paint_descendants(context, m_box, StackingContextPaintPhase::Floats);
-    // Draw inline content, replaced content, etc. (steps 6, 7)
-    paint_descendants(context, m_box, StackingContextPaintPhase::BackgroundAndBordersForInlineLevelAndReplaced);
-    paint_node(m_box, context, PaintPhase::Foreground);
-    paint_descendants(context, m_box, StackingContextPaintPhase::Foreground);
+    auto paint_box_content = [&context](Layout::Box& box) {
+        // Draw the background and borders for block-level children (step 4)
+        paint_descendants(context, box, StackingContextPaintPhase::BackgroundAndBorders);
+        // Draw the non-positioned floats (step 5)
+        paint_descendants(context, box, StackingContextPaintPhase::Floats);
+        // Draw inline content, replaced content, etc. (steps 6, 7)
+        paint_descendants(context, box, StackingContextPaintPhase::BackgroundAndBordersForInlineLevelAndReplaced);
+        paint_node(box, context, PaintPhase::Foreground);
+        paint_descendants(context, box, StackingContextPaintPhase::Foreground);
+    };
+
+    paint_box_content(m_box);
 
     // Draw positioned descendants with z-index `0` or `auto` in tree order. (step 8)
     // NOTE: Non-positioned descendants that establish stacking contexts with z-index `0` or `auto` are also painted here.
     // FIXME: There's more to this step that we have yet to understand and implement.
-    m_box.paint_box()->for_each_in_subtree_of_type<PaintableBox>([&](PaintableBox const& paint_box) {
-        auto const& z_index = paint_box.computed_values().z_index();
+    paintable().for_each_in_subtree_of_type<PaintableBox>([&](PaintableBox const& paint_box) {
+        auto z_index = paint_box.computed_values().z_index();
         if (auto* child = paint_box.stacking_context()) {
-            if (!z_index.has_value() || z_index.value() == 0)
+            if (z_index.value_or(0) == 0)
                 paint_child(child);
             return TraversalDecision::SkipChildrenAndContinue;
-        }
-        if (z_index.has_value() && z_index.value() != 0)
-            return TraversalDecision::Continue;
-        if (!paint_box.layout_box().is_positioned())
+        } else {
```

Comment:
Why use an `else` here if the upper branch always returns?

Reply:
I find if/else much easier to parse to be honest, i parse them top-level as two branches of a decision, but I am not strong about this